spyder/plugins/ipythonconsole/widgets/debugging.py

```diff
@@ -10,65 +10,40 @@
 """
 
 import re
-import pdb
 
-from IPython.core.history import HistoryManager
-from qtconsole.rich_jupyter_widget import RichJupyterWidget
-
-from spyder.config.base import get_conf_path
 from spyder.config.manager import CONF
 from spyder.py3compat import PY2
+from spyder.plugins.ipythonconsole.widgets import DebuggingHistoryWidget
 if not PY2:
     from IPython.core.inputtransformer2 import TransformerManager
 else:
     from IPython.core.inputsplitter import IPythonInputSplitter
 
 
-class PdbHistory(HistoryManager):
-
-    def _get_hist_file_name(self, profile=None):
-        """
-        Get default pdb history file name.
-
-        The profile parameter is ignored, but must exist for compatibility with
-        the parent class.
-        """
-        return get_conf_path('pdb_history.sqlite')
-
-
-class DebuggingWidget(RichJupyterWidget):
+class DebuggingWidget(DebuggingHistoryWidget):
     """
     Widget with the necessary attributes and methods to handle
     communications between a console in debugging mode and
     Spyder
     """
-    PDB_HIST_MAX = 400
 
     def __init__(self, *args, **kwargs):
-        self._pdb_in_loop = False
-        self._previous_prompt = None
+        # Communication state
+        self._pdb_in_loop = False  # Are we in a debugging loop
+        self._pdb_input_ready = False  # Can we send a command now
+        self._waiting_pdb_input = False  # Are we waiting on the user
+        # Other state
+        self._pdb_prompt = (None, None)  # prompt, password
```

Comment:
ipykernel supports a password to be able to send messages to the kernel. This means every time a prompt is sent to the frontend, a password is sent as well. I am not sure in which case this password is not None, and the password feature is not tested so it might not work, but if this is needed it is already implemented.